addons/binding/org.openhab.binding.freebox/src/main/java/org/openhab/binding/freebox/internal/FreeboxHandlerFactory.java

```diff
@@ -85,24 +114,75 @@ protected ThingHandler createHandler(Thing thing) {
     protected synchronized void removeHandler(ThingHandler thingHandler) {
         if (thingHandler instanceof FreeboxHandler) {
-            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(thingHandler.getThing().getUID());
-            if (serviceReg != null) {
-                // remove discovery service, if bridge handler is removed
-                FreeboxDiscoveryService service = (FreeboxDiscoveryService) bundleContext
-                        .getService(serviceReg.getReference());
-                service.deactivate();
-                serviceReg.unregister();
-                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
-            }
+            unregisterDiscoveryService(thingHandler.getThing());
+        } else if (thingHandler instanceof FreeboxThingHandler) {
+            unregisterAudioSink(thingHandler.getThing());
         }
         super.removeHandler(thingHandler);
     }
 
     private void registerDiscoveryService(FreeboxHandler bridgeHandler) {
         FreeboxDiscoveryService discoveryService = new FreeboxDiscoveryService(bridgeHandler);
         discoveryService.activate();
-        this.discoveryServiceRegs.put(bridgeHandler.getThing().getUID(), bundleContext
+        discoveryServiceRegs.put(bridgeHandler.getThing().getUID(), bundleContext
                 .registerService(DiscoveryService.class.getName(), discoveryService, new Hashtable<String, Object>()));
     }
 
+    private void unregisterDiscoveryService(Thing thing) {
+        ServiceRegistration<?> serviceReg = discoveryServiceRegs.get(thing.getUID());
+        if (serviceReg != null) {
+            // remove discovery service, if bridge handler is removed
+            FreeboxDiscoveryService service = (FreeboxDiscoveryService) bundleContext
+                    .getService(serviceReg.getReference());
+            service.deactivate();
+            serviceReg.unregister();
+            discoveryServiceRegs.remove(thing.getUID());
+        }
+    }
+
+    private void registerAudioSink(FreeboxThingHandler thingHandler) {
+        String callbackUrl = createCallbackUrl();
+        FreeboxAirPlayAudioSink audioSink = new FreeboxAirPlayAudioSink(thingHandler, audioHTTPServer, callbackUrl);
+        @SuppressWarnings("unchecked")
+        ServiceRegistration<AudioSink> reg = (ServiceRegistration<AudioSink>) bundleContext
+                .registerService(AudioSink.class.getName(), audioSink, new Hashtable<String, Object>());
+        audioSinkRegistrations.put(thingHandler.getThing().getUID(), reg);
+    }
+
+    private void unregisterAudioSink(Thing thing) {
+        ServiceRegistration<AudioSink> reg = audioSinkRegistrations.get(thing.getUID());
+        if (reg != null) {
+            reg.unregister();
+        }
+    }
+
+    private String createCallbackUrl() {
+        if (callbackUrl != null) {
+            return callbackUrl;
+        } else {
+            final String ipAddress = NetUtil.getLocalIpv4HostAddress();
```

Comment:
Re-use method of the Sonos binding. You will find it in most bindings implementing an audio sink.
I will correct it.